### ISSUE TITLE
Document outline does not take role="heading" and aria-level attributes into account

### ISSUE THREAD

**Issue by Dan503** [validator/validator#688]
I pasted the following html into the validator:

````html
<!DOCTYPE html>
<html lang="en">
<head>
  <title>Document outline test</title>
</head>
<body>
  <div role="heading" aria-level="1">Heading 1</div>
  <div role="heading" aria-level="2">Heading 2</div>
  <div role="heading" aria-level="3">Heading 3</div>
</body>
</html>
````

This is the document outline that I expected to see:

![The outline you would see if using h# elements](https://user-images.githubusercontent.com/10610368/43173256-465ef350-8ff7-11e8-9e48-cc222c3ad6fa.png)

This is the document outline that I got back:

![An empty document outline](https://user-images.githubusercontent.com/10610368/43173294-849be092-8ff7-11e8-8c48-f21ff0c0f3a5.png)

I then tried using `<h6>` headings and altering their level with the `aria-level` attribute:

````html
<!DOCTYPE html>
<html lang="en">
<head>
  <title>Document outline test</title>
</head>
<body>
  <h6 aria-level="1">Heading 1</h6>
  <h6 aria-level="2">Heading 2</h6>
  <h6 aria-level="3">Heading 3</h6>
</body>
</html>
````

This was the result:

![A document outline that only consists of h6 elements](https://user-images.githubusercontent.com/10610368/43173364-dba220fe-8ff7-11e8-885e-35c27e3c682b.png)

**Comment by hober** [comment]
I believe this is the correct behavior—ARIA attributes should not affect things outside of the accessibility layer.

**Comment by Dan503** [comment]
> ARIA attributes should not affect things outside of the accessibility layer.

Isn't the whole point of the document outline so that heading levels are accessible to screen readers? This is portraying the document outline incorrectly.

The whole point of `role="heading"` and `aria-level="[number]"` is so that you can make headings without needing to explicitly make an element using a <h#> element.

> The purpose of this technique is to provide a way for Assistive Technologies (AT) to identify a piece of content as a heading. Applying `role="heading"` to an element causes an AT (like a screen reader) to treat it as though it were a heading.
> https://www.w3.org/TR/WCAG20-TECHS/ARIA12.html

The most common use case for this is when you need a level 7 heading. A `<h7>` element is invalid HTML so the easiest way to apply a level 7 heading is `<h6 aria-level="7">`.

**Comment by sideshowbarker** [comment]
I don’t understand the use case for the actual examples cited in the issue description.

The first example has this:

```html
  <div role="heading" aria-level="1">Heading 1</div>
  <div role="heading" aria-level="2">Heading 2</div>
  <div role="heading" aria-level="3">Heading 3</div>
```
That seems bad to me and I don’t understand why anybody would want to do that instead of using actual h1-h3 heading elements. The h1-h6 heading elements are HTML’s native semantics for headings. Why avoid using them? As far as I understand, using a `div` as a faux heading is pretty much universally considered to be a bad-practice antipattern.

Or is there some real-world case where using h1-h6 headings creates a problem that using a `div` as a heading doesn’t?

The second example has this:

```
  <h6 aria-level="1">Heading 1</h6>
  <h6 aria-level="2">Heading 2</h6>
  <h6 aria-level="3">Heading 3</h6>
```

That example seems like just plain bad practice. Wouldn’t that cause AT users to be presented with a different heading-level hierarchy than non-AT users? Why would any author-developer ever want to actually do that to their users? The non-AT users would just see those as h6 headings, while AT users would seem those as h1-h3 headings?

**Comment by Dan503** [comment]
@sideshowbarker they were just examples to highlight that the validator isn't working correctly. I don't actually think anyone should use that sort of markup if they can avoid it.

There are two real world use cases that occur quite frequently as to why you would use `role="heading"` and `aria-level` over regular heading elements.

1. The CMS is outputting `<div>` elements instead of heading elements and there is no way to get the CMS to output proper heading elements. Instead of replacing the whole element with a heading element which is difficult and has bad performance costs, the developer adds `role="heading"` and `aria-role` attributes.

2. You have already reached a `<h6>` level heading and need to go down one more level. There is no such thing as an `<h7>` element though so using one would be as inaccessible as using a `<div>` element. So instead, the developer uses an `<h6>` element and adds `aria-level="7"`.

I have personally come across both these issues multiple times in real world websites I've worked on.

This issue isn't a case of _should_ a developer do this. This is a case of if the developer has to resort to this, the validator should reflect the real document outline structure by taking aria levels into account.

**Comment by nschonni** [comment]
http://w3c.github.io/aria/#co-evolution
> It is not appropriate to create objects with style and script when the host language provides a semantic element for that type of object. While WAI-ARIA can improve the accessibility of these objects, accessibility is best provided by allowing the user agent to handle the object natively. For example, it's better to use an h1 element in HTML than to use the heading role on a div element.

But I get that a CMS can break stuff.

**Comment by sideshowbarker** [comment]
> But I get that a CMS can break stuff.

I do too. But from that it doesn’t follow that the rest of us should change the behavior of our tools to facilitate/compensate for the use of broken CMSes

**Comment by Dan503** [comment]
> But from that it doesn’t follow that the rest of us should change the behavior of our tools to facilitate/compensate for the use of broken CMSes

Ok but don't you think that the official W3C document outline tool should abide by the the W3C standards for generating the document outline?

**Comment by sideshowbarker** [comment]
> Ok but don't you think that the official W3C document outline tool should abide by the the W3C standards for generating the document outline?

The only standard for generating a document outline is the outline algorithm in the HTML spec:

https://html.spec.whatwg.org/multipage/sections.html#outlines

…and that algorithm intentionally doesn’t take `aria-level` values into consideration. So an outline tool that claims to follow the standards but that changed its behavior based on `aria-level` would be a case of explicitly not abiding by the standards.

**Comment by sideshowbarker** [comment]
>     1. The CMS is outputting `<div>` elements instead of heading elements and there is no way to get the CMS to output proper heading elements. Instead of replacing the whole element with a heading element through JavaScript which is difficult and has bad performance costs, the developer adds `role="heading"` and `aria-role` attributes instead.

So yeah that CMS is broken. And while I genuinely empathize with developers who get stuck dealing with broken systems like that, my sense of empathy doesn’t extend to being wiling to spend my time changing the behavior of tools I develop in order to facilitate/compensate for other tools being broken.

**Comment by sideshowbarker** [comment]
> 2. You have already reached a `<h6>` level heading and need to go down one more level. There is no such thing as an `<h7>` element though so using one would be as inaccessible as using a `<div>` element. So instead, the developer uses an `<h6>` element and adds `aria-level="7"`.

So you’d end up with non-AT users getting something different than AT users. To me at least, adding `aria-level="7"` doesn’t seem like a great solution to that problem. The problem is that you’ve reached a limitation in the core HTML language itself. In my experience at least, the way that developers/authors at that point deal with that is by switching to some other HTML-native structuring method — dl/dt/dd or ol/li or something. Or by using nested `section` elements rather than just using h1-h6.

**Comment by sideshowbarker** [comment]
Closing this, since the behavior described here is intentional and per-spec. But feel free to comment further if there’s new information — and we could reopen this if needed

**Comment by Dan503** [comment]
I get the feeling that I should create an issue against the HTML spec then because it makes no sense having `<h1>` affect the document outline and `<div role="heading" aria-level="1">` not affect the document outline. They should have identical functionality (excluding styling).

**Comment by Dan503** [comment]
> switching to some other HTML-native structuring method

That isn't necessarily the most semantically correct method of representing the content though. 😕

> Or by using nested section elements

I would _love_ to use the document outline algorithm method if I could but there are no browsers that support it. There doesn't look to be any support coming on the horizon either 😢
